src/MainWindow.vala

```diff
@@ -228,14 +225,33 @@ namespace PantheonTerminal {
             destroy.connect (on_destroy);
 
             restorable_terminals = new HashTable<string, TerminalWidget> (str_hash, str_equal);
+
+            if (recreate_tabs) {
+                open_tabs ();
+            }
         }
 
         public void add_tab_with_command (string command) {
             new_tab ("", command);
         }
 
-        public void add_tab_with_working_directory (string location) {
-            new_tab (location);
+        public void add_tab_with_working_directory (string location, bool allow_duplicate = false) {
+            /* This requires all restored tabs to be initialized first so that the shell location is available */
+            bool add_tab = true;
+            if (!allow_duplicate) {
+                var f1 = File.new_for_commandline_arg (location);
+                terminals.foreach ((t) => {
```

Comment:
It seems that we should stop using `GLib.List` which forces us to use `@foreach` which doesn't support `return` or `break`. I would change `terminals` to use `Gee.ArrayList` or `Gee.LinkedList` and then immedietely return here if we found an identical tab.

Reply:
If the `foreach (TerminalWidget t in terminals {}` construction is used it is possible breakout of the loop without changing from GList and saving a local variable.   This is what I have done in the latest commit.  Changing to a Gee Collection requires more extensive and unrelated changes and should be a separate PR, I think.  It would also be possible to use a Vala.ArrayList at the expense of adding another dependency.  At some point it should be possible to replace the Gee dependency with a libvala dependency?